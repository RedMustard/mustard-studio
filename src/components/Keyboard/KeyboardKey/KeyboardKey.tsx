import { h } from 'preact';
import { useState } from 'preact/hooks';

interface KeyboardKeyProps {
    className: string;
    onClick: h.JSX.MouseEventHandler<HTMLDivElement>;
    onMouseDown: h.JSX.TargetedEvent<HTMLDivElement, Event>;
}

const KeyboardKey = ({ className, onClick, onMouseDown }: KeyboardKeyProps) => {
    const [isPressed, setIsPressed] = useState(false);

    const handleOnMouseEnter = (e: h.JSX.TargetedEvent<HTMLDivElement, Event>) => {
        setIsPressed(true);
        // onMouseDown(e.target.va);
    };
    const handleOnMouseLeave = (e: h.JSX.TargetedEvent<HTMLDivElement, Event>) => {
        setIsPressed(false);
        // onMouseDown(e.target.va);
    };

    return (
        <div
            class={`key ${className} ${isPressed ? 'key--active' : ''}`}
            label="white-key"
            onClick={onClick}
            onKeyDown={() => {}}
            onMouseEnter={(e: h.JSX.TargetedEvent<HTMLDivElement, Event>) => handleOnMouseEnter(e)}
            onMouseLeave={(e: h.JSX.TargetedEvent<HTMLDivElement, Event>) => handleOnMouseLeave(e)}
            role="button"
            tabIndex={0}
            type="button"
        />
    );
};


export default KeyboardKey;
