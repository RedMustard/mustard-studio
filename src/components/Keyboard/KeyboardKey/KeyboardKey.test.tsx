import { h } from 'preact';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { KeyboardKey } from './KeyboardKey';


const baseProps = {
    onMouseDown: jest.fn(),
    onMouseUp: jest.fn(),
    onMouseOver: jest.fn(),
    onMouseLeave: jest.fn(),
};

describe('<KeyboardKey />', () => {
    it('renders with basic props', () => {
        const wrapper = shallow(
            <KeyboardKey
                {...baseProps}
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders with custom class', () => {
        const wrapper = shallow(
            <KeyboardKey
                className="custom-class"
                {...baseProps}
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders active after onMouseOver', () => {
        const wrapper = shallow(
            <KeyboardKey
                {...baseProps}
            />,
        );
        wrapper.find('div').simulate('mouseover');
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(baseProps.onMouseOver).toBeCalled();
    });

    it('renders active after onMouseDown', () => {
        const wrapper = shallow(
            <KeyboardKey
                {...baseProps}
            />,
        );
        wrapper.find('div').simulate('mousedown');
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(baseProps.onMouseOver).toBeCalled();
        expect(baseProps.onMouseDown).toBeCalled();
    });

    it('renders inactive after onMouseOver and onMouseLeave', () => {
        const wrapper = shallow(
            <KeyboardKey
                {...baseProps}
            />,
        );
        wrapper.find('div').simulate('mouseover');
        wrapper.find('div').simulate('mouseleave');
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(baseProps.onMouseOver).toBeCalled();
        expect(baseProps.onMouseLeave).toBeCalled();
    });

    it('renders inactive after onMouseDown and onMouseUp', () => {
        const wrapper = shallow(
            <KeyboardKey
                {...baseProps}
            />,
        );
        wrapper.find('div').simulate('mousedown');
        wrapper.find('div').simulate('mouseup');
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(baseProps.onMouseOver).toBeCalled();
        expect(baseProps.onMouseDown).toBeCalled();
        expect(baseProps.onMouseUp).toBeCalled();
    });
});
