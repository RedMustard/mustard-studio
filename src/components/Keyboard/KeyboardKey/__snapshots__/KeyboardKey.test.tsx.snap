// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<KeyboardKey /> renders active after onMouseDown 1`] = `
<div
  className="keyboard-key  keyboard-key--active"
  label="keyboard-key"
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
  onMouseUp={[Function]}
  role="button"
  tabIndex={0}
  type="button"
/>
`;

exports[`<KeyboardKey /> renders active after onMouseOver 1`] = `
<div
  className="keyboard-key  keyboard-key--active"
  label="keyboard-key"
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
  onMouseUp={[Function]}
  role="button"
  tabIndex={0}
  type="button"
/>
`;

exports[`<KeyboardKey /> renders inactive after onMouseDown and onMouseUp 1`] = `
<div
  className="keyboard-key  "
  label="keyboard-key"
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
  onMouseUp={[Function]}
  role="button"
  tabIndex={0}
  type="button"
/>
`;

exports[`<KeyboardKey /> renders inactive after onMouseOver and onMouseLeave 1`] = `
<div
  className="keyboard-key  "
  label="keyboard-key"
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
  onMouseUp={[Function]}
  role="button"
  tabIndex={0}
  type="button"
/>
`;

exports[`<KeyboardKey /> renders with basic props 1`] = `
<div
  className="keyboard-key  "
  label="keyboard-key"
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
  onMouseUp={[Function]}
  role="button"
  tabIndex={0}
  type="button"
/>
`;

exports[`<KeyboardKey /> renders with custom class 1`] = `
<div
  className="keyboard-key custom-class "
  label="keyboard-key"
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
  onMouseUp={[Function]}
  role="button"
  tabIndex={0}
  type="button"
/>
`;
